File: backend/schemas/workout_plan.py
from pydantic import BaseModel, ConfigDict, Field


class PlanExerciseInput(BaseModel):
    exercise_id: int
    order_index: int = Field(..., ge=1)
    target_sets: int = Field(..., ge=1)
    target_reps: int = Field(..., ge=1)
    target_weight: float | None = Field(default=None, ge=0)


class PlanExerciseOut(BaseModel):
    id: int
    exercise_id: int
    order_index: int
    target_sets: int
    target_reps: int
    target_weight: float | None

    model_config = ConfigDict(from_attributes=True)


class WorkoutPlanCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=100)
    is_template: bool = False
    exercises: list[PlanExerciseInput] = Field(..., min_length=1)


class WorkoutPlan(BaseModel):
    id: int
    name: str
    creator_id: int
    is_template: bool
    exercises: list[PlanExerciseOut]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/schemas/workout-plan.ts
import { z } from 'zod'

export const PlanExerciseSchema = z.object({
  id: z.number(),
  exercise_id: z.number(),
  order_index: z.number(),
  target_sets: z.number(),
  target_reps: z.number(),
  target_weight: z.number().nullable(),
})
export type PlanExercise = z.infer<typeof PlanExerciseSchema>

export const PlanExerciseInputSchema = z.object({
  exercise_id: z.number(),
  order_index: z.number().min(1),
  target_sets: z.number().min(1),
  target_reps: z.number().min(1),
  target_weight: z.number().min(0).nullable().optional(),
})
export type PlanExerciseInput = z.infer<typeof PlanExerciseInputSchema>

export const WorkoutPlanSchema = z.object({
  id: z.number(),
  name: z.string(),
  creator_id: z.number(),
  is_template: z.boolean(),
  exercises: z.array(PlanExerciseSchema),
})
export type WorkoutPlan = z.infer<typeof WorkoutPlanSchema>

export const WorkoutPlanCreateSchema = z.object({
  name: z.string(),
  is_template: z.boolean().optional(),
  exercises: z.array(PlanExerciseInputSchema).min(1),
})
export type WorkoutPlanCreate = z.infer<typeof WorkoutPlanCreateSchema>
